hashing/src/chunk.rs

```diff
@@ -1,14 +1,49 @@
+use std::convert::TryFrom;
+
 use crate::{
-    error::MerkleVerificationError,
     primitives::{blake2b_hash, IndexedMerkleProof},
     MerkleConstructionError,
 };
 
+#[cfg_attr(
+    feature = "std",
+    derive(Debug, schemars::JsonSchema, serde::Serialize, serde::Deserialize),
+    schemars(with = "String", description = "Hex-encoded hash digest."),
+    serde(
+        deny_unknown_fields,
+        try_from = "ChunkWithProofDeserializeValidator<N>"
+    )
+)]
 pub struct ChunkWithProof<const N: usize> {
     proof: IndexedMerkleProof,
     chunk: Vec<u8>,
 }
 
+#[cfg_attr(
+    feature = "std",
+    derive(serde::Deserialize),
+    serde(deny_unknown_fields)
+)]
+pub struct ChunkWithProofDeserializeValidator<const N: usize> {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl<const N: usize> TryFrom<ChunkWithProofDeserializeValidator<N>> for ChunkWithProof<N> {
+    type Error = MerkleConstructionError;
+    fn try_from(value: ChunkWithProofDeserializeValidator<N>) -> Result<Self, Self::Error> {
+        let candidate = Self {
+            proof: value.proof,
+            chunk: value.chunk,
+        };
+        if candidate.is_valid() {
+            Ok(candidate)
+        } else {
+            Err(MerkleConstructionError::IncorrectChunkProof)
+        }
+    }
+}
+
 impl<const N: usize> ChunkWithProof<N> {
```

Comment:
```suggestion
impl<const N: usize> SizedChunkWithProof<N> {
```

Still not sure about a const-generic here... outside of testing in this file, a programmer should never construct a `ChunkWithProof` with `CHUNK_SIZE ≠ 1_048_576`.  How do we enforce that?

Reply:
That's a very good point, code is reverted to the initial solution which uses the #[cfg(test)] attribute.